### PR TITLE
feat!: multiple blobs per PFB

### LINKED ISSUE #388: ADR: Pay for multiple blobs with a single `MsgPayForBlob`
Currently, each `MsgPayForData` transaction can only pay for a single message. Ideally, users would be able to pay for multiple messages using a single payment transaction. Not only that, but because this reduces the number of state transitions each full node would have to execute, we would want to incentivize this behavior, by adding some base fee to `PayForData`

### LINKED ISSUE #890: Add the ability to include multiple `sdk.Msg`s in a `sdk.Tx` that contains a PFD
Currently, we limit the number of `sdk.Msg` that are allowed in a viable `sdk.Tx`. This is due in a large part because of the complexity that it would add to the creation of `MsgWirePayForData`s and the malleation process itself. 

Signature validation occurs over the entire `sdk.Tx`, so in order for this feature to work with malleation, we have to replace that signature with one that is valid for the malleated transaction that contains a `MsgPayForData`. 

There could be a few different approaches implementation wise:
- During the malleation process in `PrepareProposal`, split the PFD into its own `sdk.Tx`
- Account for the other `sdk.Msg`s in the transaction when signing commitments in a `MsgWirePayForData`
- Eliminate the malleation process entirely by refactoring commitment verification similar to #835

### PR DESCRIPTION
**Pull request by evan-forbes**
## Overview

This PR implements the ability to add an arbitrary number of blobs to a single PFB. Leaving as a draft until we merge the blocking [PR](https://github.com/celestiaorg/celestia-core/pull/922) in core, and probably tidy up a bit, rebase, or add a few more unit tests.

In part thanks to us planning ahead earlier this year, there's actually not that many required changes to get multiple blobs per PFB. Mostly just adding tests, using a **new mechanism to calculate the share commitments**, making a few things a slice instead of a single thing, and minor adjustments to make square estimation/square layout work with multiple blobs.

closes #388 

## Checklist

- [x] New and updated code has appropriate documentation
- [x] New and updated code has new and/or updated testing
- [x] Required CI checks are passing
- [x] Visual proof for any user facing features like CLI or documentation updates
- [x] Linked issues closed with keywords

### CODE REVIEW

**Comment by evan-forbes** [comment]
we should be able to split one or two minor refactors out of this PR

File: proto/blob/event.proto
Comment (rootulp) [inline proto/blob/event.proto:9]:
```suggestion
// EventPayForBlob defines an event that is emitted with the signer and the
// total shares used at the end of a pay for blob txn
```

File: proto/blob/tx.proto
Comment (rootulp) [inline proto/blob/tx.proto:37]:
```suggestion
  // share_versions are the versions of the share format that the blobs associated
  // with this message should use when included in a block. The share_versions
  // specified must match the share_versions used to generate the
  // share_commitment in this message.
```

File: testutil/blobfactory/payforblob_factory.go
Comment (rootulp) [inline testutil/blobfactory/payforblob_factory.go:256]:
```suggestion
// ManyMultiBlobTxSameSigner generates and returns many blob transactions with the
```

File: x/blob/client/cli/wirepayfordata.go
Comment (rootulp) [inline x/blob/client/cli/wirepayfordata.go:45]:
```suggestion
			// TODO: allow for more than one blob to be submitted via the cli
```

File: x/blob/types/payforblob.go
Comment (rootulp) [inline x/blob/types/payforblob.go:218]: [optional] we may want to cast the appconst instead of the shareVersion to avoid an overflow in case a shareVersion > the size of uint8 makes it here

```suggestion
		if v != uint32(appconsts.ShareVersionZero) {
```

**Review by rootulp — COMMENTED** [review]
Overall looks really good to me! Hyped for this feature

File: x/blob/client/cli/wirepayfordata.go
Comment (cmwaters) [reply x/blob/client/cli/wirepayfordata.go:45]: Also should we change the name of this file?

**Review by cmwaters — COMMENTED** [review]
Just some high level comments first

File: app/process_proposal.go
Comment (cmwaters) [inline app/process_proposal.go:132]: Why are we removing this? We should still be counting the amount of blobs that PFBs are paying for.

File: app/process_proposal.go
Comment (cmwaters) [inline app/process_proposal.go:91]: If it's not possible for a correct node to generate this in `PrepareProposal` then we should outright reject then entire proposal. Same for line 96

File: x/blob/client/cli/wirepayfordata.go
Comment (rootulp) [reply x/blob/client/cli/wirepayfordata.go:45]: https://github.com/celestiaorg/celestia-app/issues/1138

File: app/estimate_square_size_test.go
Comment (evan-forbes) [reply app/estimate_square_size_test.go:85]: yeah

File: app/process_proposal.go
Comment (evan-forbes) [reply app/process_proposal.go:91]: good point, I agree and don't see why we shouldn't change this now. e9a8014d39f49cb1b832d3272ca4f5d8ba46bdc9

I tried to test this, but it would take a ton of code to hit this test, as we'd basically have to write a custom share splitting method just to create an other wise valid block.

File: pkg/inclusion/get_commit.go
Comment (evan-forbes) [reply pkg/inclusion/get_commit.go:24]: > I have a question regarding ordering: how does a user know which order to sort their blobs prior to generating this merkle root. Blobs posted to different namespaces are guaranteed to be sorted by namespace, but what about blobs that are posted to the same namespace?

good question, the user can put them in an arbitrary order, and the proposer will match the starting index to the size in pfb.sizes

File: proto/blob/tx.proto
Comment (evan-forbes) [reply proto/blob/tx.proto:29]: I actually think there will be a lot of use cases where people post PFBs with many different namespaces, but better yet, we have https://github.com/celestiaorg/celestia-app/issues/1115

File: proto/blob/tx.proto
Comment (evan-forbes) [reply proto/blob/tx.proto:38]: there could maybe be a scenario with multiple share_versions, but we could enforce that if we wanted to. Its probably a good trade-off

File: x/blob/types/blob_tx.go
Comment (evan-forbes) [reply x/blob/types/blob_tx.go:14]: adding that method in that exact way might not work since BlobTx is in core, but we still could replace it with BlobTx I think

**Review by evan-forbes — COMMENTED** [review]
addressed some feedback here, but I still plan on breaking off one refactor in a meager attempt to reduce the size of the PR.

File: proto/blob/tx.proto
Comment (cmwaters) [reply proto/blob/tx.proto:38]: Ok perhaps a better question: What changes would cause us to change the share version?

File: app/process_proposal.go
Comment (cmwaters) [reply app/process_proposal.go:132]: > If there is an extra PFB, then the starting positions will also be off.

Are you sure? Don't PFBs use compact shares, couldn't you have multiple PFBs in a single share which means the starting position of the blobs would not be offset?

File: pkg/inclusion/get_commit.go
Comment (cmwaters) [inline pkg/inclusion/get_commit.go:24]: What's the advantage of merkelizing all the commitments over plain hashing

File: proto/blob/tx.proto
Comment (evan-forbes) [reply proto/blob/tx.proto:38]: One use case is to add the sender of the PFB, or other metadata, before the blob in the square. 

Other usecases would be to allow posters to not use varints or some other encoding should it be useful. 

I'm not sure if people will ever want to mix those tho. Even if they do, then only in that specific case do we have more linear increase in state transitions per different share version.

With the benefit of removing a lot of redundant bytes. Which sounds like a really good tradeoff.

File: app/process_proposal.go
Comment (evan-forbes) [reply app/process_proposal.go:132]: ahh, I see. I didn't word that well, by starting positions I meant the starting index of the blob that wraps the PFB.

If there's an extra PFB, then we check that the share commitment is included using the starting position that wraps it. there's no way to fake that without actually including the blob, so we're gucci

File: pkg/inclusion/get_commit.go
Comment (evan-forbes) [reply pkg/inclusion/get_commit.go:24]: good question, so if we have many blobs being paid for by a PFB, then there might be some scenario where you want to prove that someone paid for a specific blob without downloading the entirely blob. other than that, I'm not really sure. Using a hash would probably work fine

File: proto/blob/tx.proto
Comment (cmwaters) [reply proto/blob/tx.proto:38]: Yeah I don't think that constraining the share versions for blobs within a single PFB is going to be a problem for users. Worst case they just have two PFBs (for each version) and pay the extra 2 cents in gas.

File: proto/blob/tx.proto
Comment (cmwaters) [reply proto/blob/tx.proto:29]: The only problem with not encoding namespace is it means we have events without namespaces, which could be annoying for indexers. Say I want to query the most popular namespaces - this now becomes super difficult.

File: x/blob/types/blob_tx.go
Comment (cmwaters) [reply x/blob/types/blob_tx.go:14]: Ah yeah then simply `func ValidateBlobTx(tx tmproto.BlobTx, txcfg client.TxEncodingConfig) error`

File: pkg/inclusion/get_commit.go
Comment (cmwaters) [reply pkg/inclusion/get_commit.go:24]: > want to prove that someone paid for a specific blob without downloading the entirely blob

Won't you still need the blob or at least the commitment for that blob? 

(I'm generally a bit hesitant about merkelizing something without a strong use case. For example, tendermint merkelizes the header which I always felt was pointless. The headers is like 400 bytes. Just download the entire thing)

File: app/estimate_square_size_test.go
Comment (evan-forbes) [reply app/estimate_square_size_test.go:63]: ahh I see, this threw me off after you mentioned it too. This only takes a single share as its blobsize, not the number of shares that are used. so we have single single share tx, and therefore

```
|C|B|
|_|_|
```

File: go.mod
Comment (evan-forbes) [inline go.mod:189]: still need to update after we create a new release of core

File: proto/blob/tx.proto
Comment (evan-forbes) [reply proto/blob/tx.proto:27]: Definitely!! started to change this in this PR, but since this PR is already large, decided to spin it off as a follow up. https://github.com/celestiaorg/celestia-app/issues/1221

File: proto/blob/tx.proto
Comment (evan-forbes) [reply proto/blob/tx.proto:29]: left a comment in #1115 to link to this comment as its a good point

File: proto/blob/tx.proto
Comment (evan-forbes) [reply proto/blob/tx.proto:38]: left an issue in https://github.com/celestiaorg/celestia-app/issues/1223

File: x/blob/types/blob_tx.go
Comment (evan-forbes) [reply x/blob/types/blob_tx.go:14]: created a followup issue for this https://github.com/celestiaorg/celestia-app/issues/1224

**Review by evan-forbes — COMMENTED** [review]
FINALLY ready for review, but still technically blocked by a new release of core

File: pkg/inclusion/get_commit.go
Comment (evan-forbes) [reply pkg/inclusion/get_commit.go]: another good point.

it would be good to get @nashqueue 's point of view on this as well for inclusion fraud proofs. We can discuss synchronously if necessary, as we should probably make a decision soon. I'm definitely fine with changing this to simply include all of the share commitments in the PFB instead of using a secondary commitment over all of them.

Comment (evan-forbes) [reply pkg/inclusion/get_commit.go]: I'm fairly certain that we would have to end up downloading all of the share commitments for a single PFB if we're attempting to prove inclusion for a single one in both of the other options, those being taking a hash over all of the commitments, or including all commitments in the PFB.

Comment (evan-forbes) [reply pkg/inclusion/get_commit.go]: If there is ever a scenario where a single PFB pays for 1000 blobs, then I could see the secondary commitment being useful, until then, probably not.

Comment (nashqueue) [reply pkg/inclusion/get_commit.go]: I think you make a tradeoff between fraud-proof size and DA bytes. In a way, you can see it if you have multiple commitments, you only need to show that one of them is broken. If you have multiple, you need all blobs to compute the merkle hash and compare that it is wrong. If there is no limit to PFBs per Tx then the worst case fraud proof size for a single commitment fraud-proof would be O(n), with n being the number of shares in a square. This is the case if you have 1 share-sized blob per Tx. So, therefore, I would say saving all commitments is the better option here.

Comment (evan-forbes) [reply pkg/inclusion/get_commit.go]: sounds good, that's what we'll do then

File: app/estimate_square_size_test.go
Comment (rootulp) [reply app/estimate_square_size_test.go:63]: ohh I see, blobSizes is the # of bytes in the blob. [No change needed] then I expect this to behave similarly and indeed it does

```suggestion
			func() [][]int { return [][]int{{1}} },
			2, 1,
```

File: app/test/block_size_test.go
Comment (rootulp) [inline app/test/block_size_test.go:184]: [question] is this conditional ever execution? If `resp == nil` were true, I would expect the assertion on the line above to fail.

File: proto/blob/event.proto
Comment (rootulp) [reply proto/blob/event.proto:10]: I think this conversation is resolved by https://github.com/celestiaorg/celestia-app/issues/1169

File: x/blob/keeper/gas_test.go
Comment (rootulp) [inline x/blob/keeper/gas_test.go:84]: 👍

File: x/blob/types/blob_tx.go
Comment (rootulp) [inline x/blob/types/blob_tx.go:79]: [nit]
```suggestion
		// check that the metadata matches
```

File: x/blob/types/payforblob.go
Comment (rootulp) [inline x/blob/types/payforblob.go:212]: [super optional] future proofing for when we add a new share version

```suggestion
	for _, shareVersion := range shareVersions {
		if !slices.Contains(appconsts.SupportedShareVersions, uint8(shareVersion)) {
			return ErrUnsupportedShareVersion
		}
	}```

**Review by rootulp — DISMISSED** [review]
Great work! Let's :shipit:

File: x/blob/types/payforblob.go
Comment (evan-forbes) [reply x/blob/types/payforblob.go:212]: noice

File: app/test/block_size_test.go
Comment (evan-forbes) [reply app/test/block_size_test.go:184]: assert fails, but doesn't stop the test, so when we try to access the resp below it will panic

File: pkg/inclusion/get_commit.go
Comment (evan-forbes) [reply pkg/inclusion/get_commit.go]: left an issue to handle after this is merged https://github.com/celestiaorg/celestia-app/issues/1231

**Comment by evan-forbes** [comment]
there are still a lot of follow ups to this PR, but if there is no more feedback, I'll try to merge this PR monday

File: x/blob/types/payforblob_test.go
Comment (rootulp) [inline x/blob/types/payforblob_test.go:351]:
```suggestion
		{name: "invalid namespace", blob: &Blob{Data: []byte{1}, NamespaceId: appconsts.TxNamespaceID, ShareVersion: uint32(appconsts.DefaultShareVersion)}, expectError: true},
```

**Review by rootulp — APPROVED** [review]

**Review by rootulp — APPROVED** [review]

**Pull request merged**
